apply-roles

```diff
@@ -5,38 +5,49 @@
 # This will stop the script when an error is returned from any of the CLI commands
 set -e
 
-readonly BRANCH="master"
-readonly GITHUB_URI="https://raw.githubusercontent.com/cloudconformity/azure-onboarding-scripts/${BRANCH}"
 readonly CUSTOM_ROLE_NAME="Custom Role - Cloud One Conformity"
 
+# utility method to join arrays by the character passed in the first argument
+array_join() {
+  local IFS="$1"
+  shift
+  echo "$*"
+}
+
 check_for_custom_role_availability() {
   echo " - Checking if custom role definition has been created..."
   custom_role_definition_id=$(az role definition list --name "${CUSTOM_ROLE_NAME}" --query "[0].name" --output tsv)
 }
 
 create_custom_role() {
   local subscription_ids_in_active_directory=("$@")
-  local prefixed_subscription_ids
   local attempts_to_retrieve_role=0
+  local assignable_scopes
 
   echo " - Creating custom role definition..."
 
   # generate set of assignable scopes for the custom role
   for subscription in "${subscription_ids_in_active_directory[@]}"; do
     # prefix subscription id with "/subscriptions/" as is needed when added to the assignable scopes
-    prefixed_ids=("${prefixed_ids[@]}" "/subscriptions/${subscription}")
+    prefixed_ids=("${prefixed_ids[@]}" "\"/subscriptions/${subscription}\"")
   done
 
-  # convert bash array into a JSON array so can be passed to role creation
-  prefixed_subscription_ids=$(printf '%s\n' "${prefixed_ids[@]}" | jq -R . | jq -s .)
-
-  az deployment sub create \
-    --location eastus \
-    --template-uri "${GITHUB_URI}/roleDefinition/create/deploy.json" \
-    --parameters \
-    "${GITHUB_URI}/roleDefinition/create/deploy.parameters.json" \
-    roleName="${CUSTOM_ROLE_NAME}" \
-    subscriptionIds="${prefixed_subscription_ids}" > /dev/null
+  assignable_scopes=$(array_join , "${prefixed_ids[@]}")
+
+  az role definition create \
+    --role-definition "{ \
+      \"Name\": \"${CUSTOM_ROLE_NAME}\", \
+      \"Description\": \"Subscription level custom role for Cloud Conformity access.\", \
+      \"Actions\": [ \
+         \"Microsoft.AppConfiguration/configurationStores/ListKeyValue/action\", \
+         \"Microsoft.Network/networkWatchers/queryFlowLogStatus/action\", \
+         \"Microsoft.Web/sites/config/list/Action\", \
+         \"Microsoft.Storage/storageAccounts/queueServices/queues/read\" \
```

Comment:
⛏️ defining **Actions** as a separate constant could make code maintenance easier

Reply:
Yeah for sure the plan is to move this into an external json file which can also be downloaded so it can be applied manually via the Azure portal.

Would just be a matter of reading the file in and then updating the name and the `assignableScopes` part of the string. Probably not that difficult but will probably involve `sed` so that will take week to understand 😄 .

I will create a ticket for this.

Reply:
Perfect then you already have better plans for it. `sed` is really powerful text matching/processing/updating tool, I'm sure there are lots of things you can achieve with it though it has a steep learning curve in the beginning